doc/PIDHandler.md

```diff
@@ -0,0 +1,229 @@
+for one collection together. Whenever metadata is involved for ParticleIDs this
+will be the thing to use.
+
+## General considerations
+
+There is no strictly enforced coupling between the meta information that is set
+for a collection and the contents of that collection. In order for everything to
+work as expected the following assumptions have to be met
+- The collection name that is passed to the `setAlgoInfo` methods to set meta
+  information has to match the collection name used for putting collections into
+  events.
+- All elements of a `ParticleIDCollection` have the same value set for
+  `algorithmType`. Additionally, it is usually assumed that this is a unique
+  value.
+- The `ParticleIDMeta::paramNames` is assumed to match the parameters as they
+  are set in the `ParticleID` elements.
+
+Additionally there are a few usage considerations to be kept in mind
+- The `PIDHandler` can be used without any meta data about the contained
+  `ParticleID`s. However, in this case it will only be useful for inverting
+  relations and for getting `ParticleID`s that point to a given
+  `ReconstructedParticle`.
+- There is no guarantee that the meta information that is set in a `PIDHandler`
+  is consistent with the elements it knows about. Some constructors attempt to
+  construct a consistent handler, but it is always possible to set meta
+  information via `setMetaInfo`, which might be completely unrelated to the
+  actual contents of a handler.
+
+Given that there is no enforced consistency the interface and the utility
+functionality makes heavy use of
+[`std::optional`](https://en.cppreference.com/w/cpp/utility/optional) in its
+return values / types. This makes it possible to check whether the desired
+information was actually present, without having to reserve some special values
+to indicate that fact.
+
+## Usage examples
+
+**The following examples assume the following includes** (which will not be visible in the examples)
+
+```cpp
+#include <edm4hep/utils/ParticleIDUtils.h>
+#include <edm4hep/ParticleIDCollection.h>
+
+#include <podio/Frame.h>
+```
+
+**The following examples additionally assume, that there is a `metadata` and an
+`event` Frame present.** The `metadata` Frame can be obtained like this from
+standard EDM4hep files.
+
+```cpp
+#include <podio/ROOTReader.h>
+#include <podio/Frame.h>
+
+podio::Frame getMetadataFrame(const std::string& filename) {
+  podio::ROOTReader reader{};
+  reader.openFile(filename);
+  return podio::Frame(reader.readNextEntry(podio::Category::Metadata));
+}
+```
+
+Alternatively it can be created as an empty Frame and then just written using
+the `podio::Category::Metadata` category name.
+
+**Finally most of the examples assume that the desired values were found and
+simply get the `value` from the returned `std::optional` directly.** This means
+that most of the times you will see lines like this
+
+```cpp
+const auto value = pidHandler.getAlgoType("someAlgo").value();
+```
```

Comment:
The problem here is that the exception messages are very cryptic:
```
terminate called after throwing an instance of 'std::bad_optional_access'
  what():  bad optional access
zsh: IOT instruction (core dumped)  ./a.out
```
so encouraging not checking is not great. At the same time I think this part is too verbose, we don't need to teach `std::optional` here.

Reply:
Maybe I have to reword this a bit. The section here introduces the assumptions we make in the examples below. This should not be read as encouraging using `value` directly. If we do the proper checks, the examples get rather long.